src/bootstrap/builder.rs

```diff
@@ -787,6 +787,27 @@ impl<'a> Builder<'a> {
             }
         }
 
+        // If we're statically linking in an LLVM compiled for ThinLTO,
+        // configure the linker accordingly.
+        if Mode::Rustc == mode && self.config.llvm_thin_lto && !self.config.llvm_link_shared {
+            if target.contains("msvc") {
+                // Here we assume the linker is lld-link.exe.
+                rustflags.arg(&format!("-Clink-arg=/opt:lldtojobs={}", self.jobs()));
+            } else {
+                // Here we assume the linker is clang. If not, there'll be
+                // linker errors.
+                rustflags.arg("-Clink-arg=-fuse-ld=lld");
+                rustflags.arg("-Clink-arg=-flto=thin");
+
+                if self.config.llvm_optimize {
+                    rustflags.arg("-Clink-arg=-O3");
+                }
+
+                // Make LLD respect the `-j` option.
+                rustflags.arg(&format!("-Clink-arg=-Wl,--thinlto-jobs={}", self.jobs()));
+            }
+        }
```

Comment:
Do we have some reference for where these options came from (e.g., are we copying what LLVM does internally today or something like that?)? If so, would be good to note that here (or explicitly say 'these were ad-hoc created').

I presume it's fine to pass these to all rustc's as link-args and this won't interfere with incremental compilation and ThinLTO across rust code? It seems a bit odd that would just all work itself out to be honest :)

Do we set similar options wrt to the jobserver (jobs) from rustc today? Does lld support the jobserver protocol (in that case this is a bit off, and we should instead tell it to hook up to the jobserver, though that's rather hard).

Reply:
> Do we have some reference for where these options came from (e.g., are we copying what LLVM does internally today or something like that?)? If so, would be good to note that here (or explicitly say 'these were ad-hoc created').

Most of these options came from the code that was there before #57286, and they didn't have comments beyond what was here.

I'll add a note on the `-O3` flag (and apparently LLVM uses `-O2` for release-with-debuginfo builds, so I'll replicate that behavior.)

> I presume it's fine to pass these to all rustc's as link-args and this won't interfere with incremental compilation and ThinLTO across rust code? It seems a bit odd that would just all work itself out to be honest :)

Yeah, I had to map it out all in my head to get here but it seems to work.

- We should only be linking one target (librustc) in that cargo invocation.
- Since (cross-crate) ThinLTO isn't supported in dylibs, it can't interact with that.
- Crate-only ThinLTO (which is enabled by default for rlibs) shouldn't be affected by linker flags.

> Do we set similar options wrt to the jobserver (jobs) from rustc today?

It doesn't look like we do. But these flags were here before #57286, and they seem like a good idea, at least.

> Does lld support the jobserver protocol (in that case this is a bit off, and we should instead tell it to hook up to the jobserver, though that's rather hard).

Doesn't look like it.

Reply:
> We should only be linking one target (librustc) in that cargo invocation.

This is presumably 99% not true due to proc macro invocations, right? I believe we have custom logic in bootstrap which makes RUSTFLAGS apply to those as well (unlike what would normally happen, where build-time dependencies do not get RUSTFLAGS)...

Reply:
Hmm, I didn't realize RUSTFLAGS would apply to those. In that case, the assumptions about which linker is being used will have to hold for both host and target.

I don't think applying ThinLTO to the proc macro builds will break anything, but it might slow things down. lld is pretty agnostic to whether you are giving it bitcode or object files, and tries to do the right thing.

Reply:
@Mark-Simulacrum Do you think anything needs to change here, or is this PR ready to land?